CDAP-2094 Correct and update usage of "base-url"

awholegunch opened the pull request:
Running [Quick Build 5](http://builds.cask.co/browse/CDAP-DQB36-5).

Fix for https://issues.cask.co/browse/CDAP-2094.
- Changes all examples of CDAP RESTful APIs to use `/v3` instead of `<base-url>`
- Adds an explanation of `base-url` to the start of each API
- All APIs now have a similar introduction
- Added an additional CLI example
- Added tabbed-parsed-literals where `curl` commands are shown
- Standardized the use of underlines for levels in the RESTful API files
- Adjusted the order of sections in the API introduction page

Revised [HTTP RESTful APIs](http://builds.cask.co/artifact/CDAP-DQB36/shared/build-5/Docs-HTML/3.5.0-SNAPSHOT/en/reference-manual/http-restful-api/index.html)

Build 2: changes `<namespace>` to `<namespace-id>` to match other parameters, and changed the parameters names in the tables to not have leading and trailing < >.

Build 3: Modified the Sphinx custom extension to support properly-constructed JSON and escape it under Windows.

Build 4: Modified the Sphinx custom extension to support leading whitespace in literals when fabricating  Windows-equivalent commands, adjust the Security description of obtaining an Access Token, corrected examples with `curl` and `-H` options on Windows.

As Build 4 never ran, re-running as Build 5 to check.

Commit:
Correct and update usage of <base-url>.

Commit:
Remove trailing space.

wolf31o2 commented on cdap-docs/admin-manual/source/security.rst:
```diff
@@ -456,13 +450,11 @@ security components are working as expected:
-  .. tabbed-parsed-literal::
-	
-    curl -v -w"\n" -X GET -u <username>:<password> <auth-uri>
+    GET -u <username>:<password> <auth-uri>
```
Isn't it more correct to say to send Basic Authorization Headers? `-u` is a `curl` thing...

wolf31o2 commented on cdap-docs/admin-manual/source/security.rst:
```diff
@@ -456,13 +450,11 @@ security components are working as expected:
 	
-  such as (assuming an authentication server at an URI ``localhost:10009`` and that you
+  such as (assuming a CDAP authentication server at a URI ``localhost:10009`` and that you
   have defined a username:password pair such as ``cdap:realtime``):
```
Can we _not_ have `realtime` in here and use `cdap:cdap` instead?

awholegunch replied:
fixed.

wolf31o2 commented on cdap-docs/admin-manual/source/security.rst:
```diff
@@ -475,11 +467,9 @@ security components are working as expected:
-  .. tabbed-parsed-literal::
-	
-    curl -v -w"\n" -X GET <base-url>/apps -H "Authorization: Bearer <access_token>"
+    GET /v3/apps -H "Authorization: Bearer <access_token>"
```
This is mixing HTTP and curl, again.

wolf31o2 commented on cdap-docs/developers-manual/source/building-blocks/datasets/fileset.rst:
```diff
@@ -107,15 +107,24 @@ Whether they are FileSets or another type of dataset is handled by the CDAP runt
-      "dataset.counts.output.path": "monday/counts.out" }'
+.. tabbed-parsed-literal::
+
+  $ curl -w"\n" -X POST "http://example.com:10000/v3/namespaces/default/apps/FileSetExample/mapreduce/WordCount/start" \
```
Does it make sense for us to have non-required options in these? (`-w"\n"`) I know that it makes the output "cleaner" for the user, but I prefer not having optional things on the command line as it can lead to confusion.

awholegunch replied:
These examples are (in many cases) intended to be copied and pasted, so that's why I've included the `-w"\n"`. But as this particular example will not work without editing (_example.com_ needs to be replaced), I can see changing this one, and similar ones that would not normally be copied and pasted without editing.

wolf31o2 commented on cdap-docs/developers-manual/source/building-blocks/datasets/fileset.rst:
```diff
@@ -107,15 +107,24 @@ Whether they are FileSets or another type of dataset is handled by the CDAP runt
+
+  $ curl -w"\n" -X POST "http://example.com:10000/v3/namespaces/default/apps/FileSetExample/mapreduce/WordCount/start" \
+    -d "{ 'dataset.lines.input.paths':  'monday/my.txt', \
+          'dataset.counts.output.path': 'monday/counts.out' }"
```
Swap the quotes. Use single quotes for containing the `-d` option and double quotes in the JSON. Single quotes are not valid JSON (according to the standard) even though many tools accept it as such.

http://stackoverflow.com/questions/14355655/jquery-parsejson-single-quote-vs-double-quote

awholegunch replied:
The problem is that it breaks on Windows. Yes, I **know** it's non-standard, but otherwise _all our Windows examples fail._

One alternative is to do some work in our extension, and **only** on Windows do the swap. I think the pattern is consistent enough to be able to do that. Would that be acceptable?

wolf31o2 replied:
What breaks on Windows? Using double quotes within the JSON?

awholegunch replied:
The outer quotes need to be double quotes. And then you need to escape all the inner double-quotes:

```
-d "'{ \"dataset.lines.input.paths\":  \"monday/my.txt\", ^
           \"dataset.counts.output.path\": \"monday/counts.out\" }'"
```

To be honest, I haven't tested this for a bit, but I think this is what's needed.

awholegunch replied:
I guess I could look at doing this just for Windows. Ugly, but that's what people get for using Windows...

awholegunch replied:
OK, this is what Windows wants: no single quotes. I'm not sure that the continuation character actually can be where I showed it, in the middle of the JSON:

```
-d "{ \"dataset.lines.input.paths\":  \"monday/my.txt\", ^
           \"dataset.counts.output.path\": \"monday/counts.out\" }"
```

wolf31o2 replied:
IMO, this is the correct way to display this on Windows, due to how it handles single/double quotes. Yeah, it's ugly. However, the original way is showing non-standard JSON and won't lint/validate in many tools/sites.

wolf31o2 commented on cdap-docs/developers-manual/source/data-exploration/tables.rst:
```diff
@@ -69,25 +69,28 @@ Setting a Schema on an Existing Table
+            }', \
+            'schema.row.field': 'id' \
+          } \
+        }"
```
Same comment re: single/double quotes in JSON

awholegunch replied:
See above.

Commit:
Edits per comments.

awholegunch commented:
One other change to consider: we use the syntax of enclosing parameters in less-than and greater-than symbols, such as:

```
PUT /v3/namespaces/<namespace>/streams/<new-stream-id>
```

whereas there are numerous RESTful APIs using curly braces. Here's an example from https://www.twilio.com/docs/api/rest/outgoing-caller-ids

```
/2010-04-01/Accounts/{AccountSid}/OutgoingCallerIds/{OutgoingCallerIdSid}
```

It might be easier to read and more consistent. Should we change to this?:

```
PUT /v3/namespaces/{namespace}/streams/{new-stream-id}
```

**NOTE:** _I decided to not attempt this, as it interferes with the JSON examples, as they use curled braces extensively._

awholegunch commented:
"On-hold" as I am going to take another pass and clean up some of the descriptions of parameters.

anew commented on cdap-docs/admin-manual/source/security.rst:
```diff
@@ -450,16 +449,17 @@ will use ``https`` instead of ``http``.
+  obtain an ``AccessToken``::
 
-    GET -u <username>:<password> <auth-uri>
+    GET <auth-uri> <username>:<password>
```
don't you still need -u here?

awholegunch replied:
You do for the HTTP curl command, but this is (intended) to be the API, and not the command...there doesn't seem to be a standard way of describing this except as:

```
<http-method> <uri> <options, headers, body>
```

anew replied:
I don't think we need the curl at all to define a REST endpoint. curl is only one way in which people make REST calls. We should describe exactly what you suggest: the HTTP method, the URL, the accepted headers and the expected request body.

wolf31o2 replied:
I think the problem is that we're trying to simplify the documentation with regard to authentication. It's much easier to `-u username:password` than to explain the encoded authorization header those options create. Personally, I think that we should just describe the header and how it's constructed, since we only need it in this one place.

https://en.wikipedia.org/wiki/Basic_access_authentication#Client_side

After all, we describe the header itself for our token. We should do the same for fetching the token in the first place.

Commit:
Strip opening and closing brackets off of parameter names in tables.

Commit:
Change all uses of "<namespace>" to "<namespace-id>" to match other parameters.

Commit:
Merge branch 'develop' into docs/3.5_2094_Base_URL

# Conflicts:
#	cdap-docs/reference-manual/source/http-restful-api/artifact.rst

Edited conflicting line to merge to the correct result.

Commit:
Revise all JSON examples to be properly-constructed. Change the
tabbed-parsed literal Sphinx extension to handle converting to an
escaped form that works under Windows.

wolf31o2 commented on cdap-docs/admin-manual/source/security.rst:
```diff
@@ -456,30 +449,27 @@ security components are working as expected:
 
-- Submit a username and password to one of the URIs to obtain an ``AccessToken``:
+- Submit a username and password to one of the authentication URIs (``<auth-uri>``) to
+  obtain an ``AccessToken``::
```
So this is incorrect. You submit a Basic Authorization header. The API doesn't take username and password.

wolf31o2 commented on cdap-docs/admin-manual/source/security.rst:
```diff
@@ -456,30 +449,27 @@ security components are working as expected:
-  .. tabbed-parsed-literal::
-	
-    curl -v -w"\n" -X GET -u <username>:<password> <auth-uri>
+    GET <auth-uri> <username>:<password>
```
This should be something more like `GET <auth-uri> "Authorization: Basic abcd1234=="` with an explanation on how to create the encoded username:password pair... "or you can use curl like..." below, with a username:password pair and it'll generate the header for you.

wolf31o2 commented on cdap-docs/developers-manual/source/building-blocks/plugins.rst:
```diff
@@ -336,7 +336,7 @@ it needs to be set explicitly, as the JAR contents are uploaded without the file
 
   $ curl -w"\n" -X POST "localhost:10000/v3/namespaces/default/artifacts/mysql-connector-java" \
-  -H "Artifact-Plugins: [ { 'name': 'mysql', 'type': 'jdbc', 'className': 'com.mysql.jdbc.Driver' } ]" \
+  -H 'Artifact-Plugins: [ { "name": "mysql", "type": "jdbc", "className": "com.mysql.jdbc.Driver" } ]' \
```
This and 129 above aren't consistent. Is this on purpose?

awholegunch replied:
Nope. That's an error. Good catch.

awholegunch replied:
Fixed. And added support for leading whitespace, and fixed a few other examples I noticed.

Commit:
Fix the tabbed parsed literal to handle leading spaces and headers.
Fixed and incorrect example of an X-App-Config.
Corrected quotes on other examples.

Commit:
Correct the description of obtaining an Access Token per comments.

wolf31o2 commented:
:+1: LGTM

awholegunch commented:
Waiting for [Build 5](http://builds.cask.co/browse/CDAP-DQB36-5) to pass before merging.

Commit:
Move trailing options to their own list, and added missing conversion of
replacing single quotes with double quotes.

awholegunch commented:
Found an error: some trailing single quotes were not being converted to double quotes correctly. Fixed and re-running build. Will merge if passes and proves correct on inspection of HTML output.
